torch/csrc/jit/codegen/cuda/executor.cpp

```diff
@@ -283,42 +285,111 @@ std::vector<at::Tensor> FusionExecutor::allocOutputs(EvaluationContext& ec) {
 std::vector<at::Tensor> FusionExecutor::runFusion(
     const at::ArrayRef<IValue>& inputs,
     const std::vector<at::Tensor>& outputs,
-    const LaunchParams& launch_constraints) {
+    const LaunchParams& launch_constraints,
+    const c10::optional<size_t>& opt_code) {
   TORCH_INTERNAL_ASSERT(
       fusion_id_ > 0, "Cannot run fusion, it was not compiled.");
+  TORCH_INTERNAL_ASSERT(
+      !opt_code.has_value() || outputs.empty(),
+      "short cut input cache is not compatible with pre-allocated output");
 
-  FusionGuard fg(&fusion_);
-
-  executor_utils::validateKernelInputs(&fusion_, inputs, options_.device);
+  ExecutorEntry* executor_entry = nullptr;
+  if (opt_code.has_value()) {
+    executor_entry = &executor_entry_lookup_[*opt_code];
+  }
 
+  FusionGuard fg(&fusion_);
   c10::DeviceGuard dg(options_.device);
   auto stream = at::cuda::getCurrentCUDAStream();
 
-  EvaluationContext evaluation_context =
```

Comment:
https://github.com/csarofeen/pytorch/pull/347
makes the evaluation of a fusion stateful. I'm wondering if we should cache the stateful evaluator here instead of everything. Maybe this introduces the chance of regressing though, and we should just move forward with the way you're already caching in this PR.

Reply:
I don't remember when we measure the number, if stateful evaluation still gives high latency. I though the number we have are pretty good.